src/sentry/taskworker/worker.py

```diff
@@ -0,0 +1,208 @@
+            logger.info("get_task failed. Retrying in 1 second")
+            return None
+
+        if not activation:
+            logger.info("No task fetched")
+            return None
+
+        return activation
+
+    def _known_task(self, activation: TaskActivation) -> bool:
+        if not taskregistry.contains(activation.namespace):
+            logger.error(
+                "taskworker.invalid_namespace",
+                extra={"namespace": activation.namespace, "taskname": activation.taskname},
+            )
+            return False
+
+        namespace = taskregistry.get(activation.namespace)
+        if not namespace.contains(activation.taskname):
+            logger.error(
+                "taskworker.invalid_taskname",
+                extra={"namespace": activation.namespace, "taskname": activation.taskname},
+            )
+            return False
+        return True
+
+    def process_task(self, activation: TaskActivation) -> TaskActivation | None:
+        assert self._pool
+        if not self._known_task(activation):
+            self._execution_count += 1
+            return self.client.update_task(
+                task_id=activation.id,
+                status=TASK_ACTIVATION_STATUS_FAILURE,
+            )
+
+        # TODO(taskworker): Add at_most_once checks
+        processing_timeout = activation.processing_deadline_duration
+        namespace = taskregistry.get(activation.namespace)
+        next_state = TASK_ACTIVATION_STATUS_FAILURE
+        result = None
+        execution_start_time = 0.0
+        try:
+            task_data_parameters = orjson.loads(activation.parameters)
+            execution_start_time = time.time()
+
+            result = self._pool.apply_async(
+                func=_process_activation,
+                args=(
+                    activation.namespace,
+                    activation.taskname,
+                    task_data_parameters["args"],
+                    task_data_parameters["kwargs"],
+                ),
+            )
+            # Will trigger a TimeoutError if the task execution runs long
+            result.get(timeout=processing_timeout)
+
+            next_state = TASK_ACTIVATION_STATUS_COMPLETE
+        except TimeoutError:
+            logger.info(
+                "taskworker.task_execution_timeout",
+                extra={
+                    "taskname": activation.taskname,
+                    "processing_deadline": processing_timeout,
+                },
+            )
+            # When a task times out we kill the entire pool. This is necessary because
+            # stdlib multiprocessing.Pool doesn't expose ways to terminate individual tasks.
+            self._build_pool()
+        except Exception as err:
+            if namespace.get(activation.taskname).should_retry(activation.retry_state, err):
+                logger.info("taskworker.task.retry", extra={"task": activation.taskname})
+                next_state = TASK_ACTIVATION_STATUS_RETRY
```

Comment:
It feels odd that the worker is deciding if a task should be retried or not. Shouldn't the worker just report a failure, and the consumer upkeep would decide to retry it or not?

Reply:
Retry policies include allow/block lists for exception types that have to be evaluated in the worker. The workers make a suggestion to the consumer that this task should be retried *if possible*. The consumer will make a final decision based on the number of retries a task has used and how many are available.